File: dart/exercises-(in_kotlin)/exercise2_network_state.kt
// ============================================================
// SE 3242: Android Application Development
// Week 2 - Kotlin Exercises
// Exercise 2: Network Request State with Sealed Classes
// Student: Precious [Last Name]
// Date: March 2026
// Description: Demonstrates sealed classes and exhaustive
//              when expressions in Kotlin, modelling the
//              possible states of a network request.
// ============================================================

/**
 * Sealed class representing all possible states of a network request.
 *
 * Because the class is sealed, the compiler knows every possible subclass,
 * which allows `when` expressions to be exhaustive without an `else` branch.
 */
sealed class NetworkState {

    /**
     * Indicates that the network request is currently in progress.
     * Declared as an `object` because there is only ever one "loading" state —
     * no extra data is needed.
     */
    object Loading : NetworkState()

    /**
     * Indicates that the network request completed successfully.
     *
     * @property data The payload returned by the request.
     */
    data class Success(val data: String) : NetworkState()

    /**
     * Indicates that the network request failed.
     *
     * @property message A human-readable description of what went wrong.
     */
    data class NetworkError(val message: String) : NetworkState()
}

/**
 * Handles a [NetworkState] by printing a human-readable message to the console.
 *
 * The `when` expression is exhaustive: because [NetworkState] is sealed the
 * compiler guarantees that every subclass is covered, so no `else` branch is
 * required.
 *
 * @param state The current state of the network request to handle.
 */
fun handleState(state: NetworkState) {
    when (state) {
        is NetworkState.Loading      -> println("⏳ Loading... please wait.")
        is NetworkState.Success      -> println("✅ Success! Data: ${state.data}")
        is NetworkState.NetworkError -> println("❌ Error: ${state.message}")
    }
}

/**
 * Simulates a network fetch by returning a [NetworkState] based on a numeric scenario.
 *
 * | scenario | returned state                              |
 * |----------|---------------------------------------------|
 * | 1        | [NetworkState.Loading]                      |
 * | 2        | [NetworkState.Success] with a sample payload |
 * | 3        | [NetworkState.NetworkError] — timeout        |
 * | other    | [NetworkState.NetworkError] — unknown error  |
 *
 * @param scenario An integer code selecting which state to simulate.
 * @return The [NetworkState] that corresponds to the given scenario.
 */
fun simulateFetch(scenario: Int): NetworkState = when (scenario) {
    1    -> NetworkState.Loading
    2    -> NetworkState.Success("User profile loaded")
    3    -> NetworkState.NetworkError("Connection timed out")
    else -> NetworkState.NetworkError("Unknown error")
}

/**
 * Entry point for Exercise 2.
 *
 * Demonstrates:
 * - Creating a list of mixed [NetworkState] instances
 * - Using `forEach` with [handleState] to process each state
 * - Using [simulateFetch] inside a `for` loop to simulate scenarios 1-3
 * - Using `filterIsInstance<NetworkState.Success>()` to extract only successful states
 */
fun main() {
    // ── Part 1: Predefined list of states ────────────────────────────────────
    println("=== Part 1: Handling a list of network states ===\n")

    val states: List<NetworkState> = listOf(
        NetworkState.Loading,
        NetworkState.Success("User profile loaded"),
        NetworkState.NetworkError("Connection timed out"),
        NetworkState.Success("Settings retrieved"),
        NetworkState.NetworkError("404 Not Found"),
        NetworkState.Loading,
        NetworkState.Success("Notifications fetched")
    )

    states.forEach { state -> handleState(state) }

    // ── Separator ─────────────────────────────────────────────────────────────
    println()
    println("------------------------------------------------------------")
    println()

    // ── Part 2: simulateFetch loop for scenarios 1–3 ─────────────────────────
    println("=== Part 2: Simulated fetch scenarios ===\n")

    for (scenario in 1..3) {
        val result = simulateFetch(scenario)
        print("Scenario $scenario → ")
        handleState(result)
    }

    println()
    println("------------------------------------------------------------")
    println()

    // ── Part 3: Filter only successful states ────────────────────────────────
    println("=== Part 3: Successful states only ===\n")

    val successfulStates = states.filterIsInstance<NetworkState.Success>()

    if (successfulStates.isEmpty()) {
        println("No successful states found.")
    } else {
        println("Found ${successfulStates.size} successful state(s):")
        successfulStates.forEach { success ->
            println("  → ${success.data}")
        }
    }

    println()
    println("Exercise 2 complete.")
}

File: dart/exercises-(in_dart)/exercise2_network_state.dart
// ============================================================
// SE 3242: Android Application Development
// Week 2 - Dart Exercises
// Exercise 2: Network Request State with Sealed Classes
// Student: [Your Name]
// Date: March 2026
// Description: Demonstrates sealed classes and exhaustive
//              pattern matching in Dart, modelling the three
//              possible states of a network request.
// ============================================================

// ─────────────────────────────────────────────────────────────
// Sealed class hierarchy
// ─────────────────────────────────────────────────────────────

/// Sealed base class representing all possible states of a
/// network request.
///
/// Because this class is `sealed`, the Dart compiler knows every
/// subclass at compile time and can enforce exhaustive pattern
/// matching in switch expressions — no `default` branch needed.
sealed class NetworkState {}

/// Represents a network request that is currently in progress.
///
/// No data is available yet while in this state.
class Loading extends NetworkState {}

/// Represents a successfully completed network request.
///
/// [data] contains the payload returned by the server.
class Success extends NetworkState {
  /// The data payload returned by the successful request.
  final String data;

  /// Creates a [Success] state with the given [data].
  Success(this.data);
}

/// Represents a network request that has failed.
///
/// [message] describes the reason for the failure.
class NetworkError extends NetworkState {
  /// A human-readable description of the error.
  final String message;

  /// Creates a [NetworkError] state with the given [message].
  NetworkError(this.message);
}

// ─────────────────────────────────────────────────────────────
// State handler
// ─────────────────────────────────────────────────────────────

/// Handles [state] by printing a descriptive message for each
/// possible [NetworkState] subtype.
///
/// Uses an exhaustive `switch` expression — because [NetworkState]
/// is `sealed`, the compiler guarantees every case is covered
/// without a `default` branch.
void handleState(NetworkState state) {
  final message = switch (state) {
    Loading()       => '⏳ Loading... please wait.',
    Success(:final data)          => '✅ Success! Data: $data',
    NetworkError(:final message)  => '❌ Error: $message',
  };
  print(message);
}

// ─────────────────────────────────────────────────────────────
// Simulated fetch
// ─────────────────────────────────────────────────────────────

/// Simulates a network fetch by returning a [NetworkState] based
/// on the given [scenario] number.
///
/// | scenario | returned state                          |
/// |----------|-----------------------------------------|
/// | 1        | [Loading]                               |
/// | 2        | [Success] with `"User profile loaded"`  |
/// | 3        | [NetworkError] with `"Connection timed out"` |
/// | other    | [NetworkError] with `"Unknown error"`   |
NetworkState simulateFetch(int scenario) {
  return switch (scenario) {
    1 => Loading(),
    2 => Success('User profile loaded'),
    3 => NetworkError('Connection timed out'),
    _ => NetworkError('Unknown error'),
  };
}

// ─────────────────────────────────────────────────────────────
// Entry point
// ─────────────────────────────────────────────────────────────

/// Demonstrates sealed classes and exhaustive pattern matching
/// through a series of simulated network state operations.
void main() {
  final separator = '─' * 50;

  // ── Section 1: forEach over a mixed list of states ────────
  print(separator);
  print('Section 1 — Handle a list of states (forEach)');
  print(separator);

  final List<NetworkState> states = [
    Loading(),
    Success('Dashboard data loaded'),
    NetworkError('404 Not Found'),
    Success('Settings saved'),
    NetworkError('No internet connection'),
  ];

  states.forEach(handleState);

  print('');

  // ── Section 2: simulateFetch for scenarios 1–3 ────────────
  print(separator);
  print('Section 2 — Simulate fetch for scenarios 1 – 3');
  print(separator);

  for (var scenario = 1; scenario <= 3; scenario++) {
    final state = simulateFetch(scenario);
    print('Scenario $scenario → ', );
    handleState(state);
  }

  print('');

  // ── Section 3: whereType<Success>() ───────────────────────
  print(separator);
  print('Section 3 — Successful states only (whereType<Success>)');
  print(separator);

  states
      .whereType<Success>()
      .forEach((s) => print('✅ Received: ${s.data}'));

  print(separator);
}
